src/fastapibootcamp/dependencies/requestcontext.py

```diff
@@ -0,0 +1,94 @@
+"""Request context FastAPI dependency."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Annotated
+
+from fastapi import Depends, Request, Response
+from httpx import AsyncClient
+from safir.dependencies.http_client import http_client_dependency
+from safir.dependencies.logger import logger_dependency
+from structlog.stdlib import BoundLogger
+
+from ..factory import Factory
+
+# This RequestContext dataclass is what will be provided by the
+# context_dependency FastAPI dependency. This is a useful pattern that wraps
+# up multiple FastAPI dependencies into one container, and is a useful
+# place to put methods that work on those dependencies (e.g., like
+# binding context to the logger.
+#
+# The RequestContext would hold the sqlalchemy Session for apps that use
+# databases, or a Redis client, etc.
+
+
+@dataclass
+class RequestContext:
+    """Holds the incoming request and its surrounding context.
+
+    The primary reason for the existence of this class is to allow the
+    functions involved in request processing to repeatedly rebind the request
+    logger to include more information, without having to pass both the
+    request and the logger separately to every function.
+    """
+
+    request: Request
+    """The incoming request."""
+
+    response: Response
+    """The response to be returned.
+
+    The response can be modified to include additional headers or status codes.
+    """
+
+    logger: BoundLogger
+    """The request logger, rebound with discovered context."""
+
+    http_client: AsyncClient
+    """The HTTPX async HTTP client."""
+
+    factory: Factory
+    """The component factory."""
+
+    def get_request_username(self) -> str | None:
+        """Get the username who made the request.
+
+        Uses the X-Auth-Request-Username header passed by Gafaelfawr.
+        """
+        return self.request.headers.get("X-Auth-Request-User")
```

Comment:
I usually use `auth_dependency` provided by Safir instead, since it fails if there is no user. Maybe the problem there is that this makes it annoying to use the app locally where there's no Gafaelfawr? (It also requires passing in a user when running tests, but that's pretty easy to handle in a fixture.)

Reply:
I must have copied this from somewhere. The auth_dependency would be a good idea for a real app. I do worry it'd make it harder to demo the API interactively locally if it fails if there are no Gafaelfawr headers.